### ISSUE TITLE
Extending to panel data

### ISSUE THREAD

**Issue by kforeman** [jessegrabowski/pymc_statespace#4]
Been watching this for a while now, and really excited to see the progress on merging it into [pymc_experimental](https://github.com/pymc-devs/pymc-experimental)! 

I've got a Bayesian hierarchical linear time series model in PyMC that I'd like to reimplement as a state space model (primarily for computational efficiency), and was poking around how to do it using your [STS](https://github.com/pymc-devs/pymc-experimental/blob/a3b0958c84ae83f4bb4cbc49f48f241242fb35ad/notebooks/Structural%20Timeseries%20Modeling.ipynb) functionality. It should be pretty trivial to port over a univariate timeseries model, but I'm curious if you have any thoughts on how best to implement for multiple timeseries with a joint likelihood.

For instance, I have some dummy covariates for specific holidays which change dates across years. Since some of the series are very short and/or noisy, I'd like to pool that information across them. Right now I have a global effect (e.g. a single `Uniform()` on `beta_global`), then add local deviations to it (e.g. `Normal(..., size=N_series)` for `beta_series`. 

Perhaps I could do something like
```
for i in range(N_series):
    mod = st...
    mod.build(name=f'ssm_{i}')
```
then create a deterministic inside the pymc model block for each series, but I'm wondering if you've already thought about this sort of use case and have better suggestions.

Also, wasn't sure where best to put this discussion - please just let me know if you'd rather have it in the PyMC Discourse or the pymc_experimental PR, or what!

**Comment by jessegrabowski** [comment]
Multivariate structural models are very very high on my to-do list. Something like what you suggest will work, but I think it would probably be quite slow, because computing the likelihood for the models is already pretty taxing. I think the best way to do it is to expand the state space and/or observation equations to include them. Brief introduction via lecture slides [here](https://personal.vu.nl/s.j.koopman/documents/2011TSEweek1.pdf), full book treatment [here](https://personal.vu.nl/s.j.koopman/old/publications/multi_stsm.pdf). 

Basically my plan is to add a `k_endog` and `endog_names` argument to every structural component, so you'll be allowed to create blocks of components for the panel case. The trickier part is the covariance matrices, I guess a keyword argument like `model_full_covariance=True` that would require you to pass a `(n_shocks, n_shocks)` matrix, then it's up to you to figure out how to structure it.

**Comment by kforeman** [comment]
Thanks, I figured you probably already had something like this in mind after seeing [this comment](https://github.com/pymc-devs/pymc-experimental/blob/a3b0958c84ae83f4bb4cbc49f48f241242fb35ad/pymc_experimental/statespace/models/structural.py#L317-L319) about `k_endog` _currently_ having to be 1 :). 

What you describe from Durbin-Koopman is also basically what others on my team have implemented for a different model within Numpyro - but I personally am much more comfortable working in PyMC, so happy to hear we'll have something to build off of here! 

If there's any roadmap items you could use some help with, will try to contribute as well.

**Comment by jessegrabowski** [comment]
I merged the statespace module into `pymc-experimental`, and started a roadmap here https://github.com/pymc-devs/pymc-experimental/issues/230, so we can move the conversation over there. I'd be happy for feedback/bugs/suggestions if you try it out!
